crates/ethereum-rpc/src/subscription.rs

```diff
@@ -1,121 +1,116 @@
+            new_heads_tx.clone(),
+            logs_tx.clone(),
+        ));
 
-                if logs_tx.receiver_count() != 0 {
-                    let mut working_set = WorkingSet::<C>::new(storage.clone());
-                    let logs = evm
-                        .get_logs_in_block_range(
-                            &mut working_set,
-                            &Filter::default(),
-                            height,
-                            height,
-                        )
-                        .expect("Error getting logs in block range");
-
-                    // Only possible error is no receiver
-                    let _ = logs_tx.send(logs);
-                }
-            }
-        });
+        let logs_notifier_handle = tokio::spawn(logs_notifier(logs_rx, logs_subscriptions.clone()));
+        let heads_notifier_handle =
+            tokio::spawn(new_heads_notifier(new_heads_rx, head_subscriptions.clone()));
 
-        manager
+        Self {
+            soft_confirmation_handle,
+            logs_notifier_handle,
+            heads_notifier_handle,
+            head_subscriptions,
+            logs_subscriptions,
+        }
     }
 
-    pub(crate) fn subscribe_new_heads(&self) -> broadcast::Receiver<RichBlock> {
-        self.new_heads_tx.subscribe()
+    pub async fn register_new_heads_subscription(&self, subscription: SubscriptionSink) {
+        let mut head_subscriptions = self.head_subscriptions.write().await;
+        head_subscriptions.retain(|s| !s.is_closed());
+        head_subscriptions.push(subscription);
     }
 
-    pub(crate) async fn subscribe_logs(&self) -> broadcast::Receiver<Vec<LogResponse>> {
-        self.logs_tx.subscribe()
+    pub async fn register_new_logs_subscription(
+        &self,
+        filter: Filter,
+        subscription: SubscriptionSink,
+    ) {
+        let mut logs_subscriptions = self.logs_subscriptions.write().await;
+        logs_subscriptions.retain(|(_, s)| !s.is_closed());
+        logs_subscriptions.push((filter, subscription));
     }
 }
 
-pub async fn handle_new_heads_subscription(
-    subscription: SubscriptionSink,
-    mut rx: broadcast::Receiver<RichBlock>,
-) {
-    tokio::spawn(async move {
-        loop {
-            let Ok(block) = rx.recv().await else {
-                // Connection closed
-                return;
-            };
+impl Drop for SubscriptionManager {
+    fn drop(&mut self) {
+        self.soft_confirmation_handle.abort();
+        self.logs_notifier_handle.abort();
+        self.heads_notifier_handle.abort();
+    }
+}
 
+pub async fn new_heads_notifier(
+    mut rx: mpsc::Receiver<RichBlock>,
+    head_subscriptions: Arc<RwLock<Vec<SubscriptionSink>>>,
+) {
+    while let Some(block) = rx.recv().await {
+        // Acquire the read lock here to prevent starving the writes.
+        let subscriptions = head_subscriptions.read().await;
+        let mut send_tasks = vec![];
+        for subscription in subscriptions.iter() {
```

Comment:
This is literally the job of broadcast.

Reply:
A subscription has type `SubscriptionSink` it's not a channel.